### ISSUE TITLE
Using argparse to parse arguments and supporting within-Python usage

### ISSUE THREAD

**Issue by tsalo** [bbfrederick/rapidtide#2]
I was thinking that it might be beneficial to use [argparse](https://docs.python.org/3/library/argparse.html) for managing the command-line arguments.

Argparse automates the construction of the help message that is currently specified manually in the `usage` function, which will make it easier to update arguments and their help info. It's also possible to manage a lot of the logic in the argument specifications, like identifying mutually exclusive arguments or limiting arguments to specific choices. The last benefit (and one that I'm particularly excited about) is that it might make it easier to compile the arguments and feed those into a separate function, which could in turn make it easier to support pure Python usage of the package. If `rapidtide` can be used as a Python module, then it will be all the easier to use it in `nipype` pipelines or to incorporate it into existing BIDS Apps (if that's something you're interested in).

I've started working on some parsing code, where I've been trying to integrate the logic that is applied later on in the `rapidtide2` code into the argument parsing [here](https://github.com/tsalo/rapidtide/blob/5833018e224e4164f2d983cec5d8899fbf02688f/rapidtide/parse.py#L29). Is this something you'd be interested in integrating, once it's complete and tested?

**Comment by bbfrederick** [comment]
Yes, definitely!   I’d thought about doing that, as the argument list is really unwieldily and hard to maintain, but I wasn’t sure how to get the formatting of the usage statement right so I delayed thinking about it. 

Making rapid ride easier to call programmatically sounds like the right direction. I’m hoping it could eventually get gilded into something like fmriprep.

**Comment by tsalo** [comment]
That's great to hear. I have the argument parser mostly done, although it looks like it will still need some post-parser processing to replicate all of the current functionality. Once I've got it figured out, I can open a PR and you can review any proposed changes.

Here are a couple of things I noticed about the arguments, which might be worth discussing here:
- `--multiproc` seems redundant with `--nprocs`. It sets the number of processes to -1, but `nprocs` could just be called with -1.
- `--datatstep` and `-t` seem to be redundant with one another. I can just set them as the same argument.
-  `-i` appears unused. I think it might have been changed to `-y`, but the usage wasn't updated.
- There are a few sets of arguments that are currently different flags, but which could be different options with the same flag (which would simplify the documentation and usage):
    - `--phat`, `--liang`, and `--eckart` (which set `corrweighting`)
    - `--ica`, `--pca`, `--weightedavg`, and `--avg` (which set `refinetype`)
    - `-V`, `-L`, `-R`, and `-C` (which set `filtertype`)
- `-h` (to set the histogram length) is duplicated with the flag to show the usage. I was thinking it could be replaced with `--histlen`.
- The variable `check_autocorrelation` is set to `True` by several arguments, but it doesn't look like it can ever be `False`, so I think it could be removed.
- The variable `shiftall` also seems to always be `True`.

**Comment by bbfrederick** [comment]
> On May 9, 2018, at 7:39 AM, Taylor Salo <notifications@github.com> wrote:
> 
> That's great to hear. I have the argument parser mostly done, although it looks like it will still need some post-parser processing to replicate all of the current functionality. Once I've got it figured out, I can open a PR and you can review any proposed changes.
> 
> Here are a couple of things I noticed about the arguments, which might be worth discussing here:
> 
> 
NB: As you read this you’ll see several options listed where double hyphen (--) has been replace by an em-dash (—).  This is due to apple mail’s aggressive and irritating spell checker.  I can fix it by laboriously cutting and pasting, but in some cases I’ve given up.  Don’t cut and paste text from this email into rapidtide command lines, though, or you’ll spend a long time chasing very hard to find bugs…

Your observations are correct.  The options with a single hyphen are the oldest ones - they come from rapidtide (no 2).  A major change to the command line syntax was the reason for rapidtide2 (rapidtide used to exist in parallel - the underlying code was the same, but the call syntax depended on the name you used to invoke the program.  That was a while ago, and for the most part, I went over to new style arguments, but a few of them were used in a LOT of my processing scripts and I didn’t want to break compatibility.)  So all new options are with the double hyphens, and I’ve been slowly deprecating the old ones.  Might be time to just bite the bullet and clean everything up though.
> --multiproc seems redundant with --nprocs. It sets the number of processes to -1, but nprocs could just be called with -1.
Yes.  --multiproc was the original option, --nprocs was added a few months ago to give finer control over how many processors are used.  multiproc is now redundant, and can go away.
> --datatstep and -t seem to be redundant with one another. I can just set them as the same argument.
Yes, -t is the old option and can be removed.

> -i appears unused. I think it might have been changed to -y, but the usage wasn't updated.
You are correct - good catch.  This could be changed to --interptype={cubic, quadratic, univariate}, univariate by default, which fixes that and gets rid of a single hyphen option.
> There are a few sets of arguments that are currently different flags, but which could be different options with the same flag (which would simplify the documentation and usage):
> --phat, --liang, and --eckart (which set corrweighting)
So --corrweighting={none, phat, liang, eckart}, none by default?  Works for me.
> --ica, --pca, --weightedavg, and --avg (which set refinetype)
Same: —refinetype={ica, pca, weightedavg, avg}, pca by default is good.
> -V, -L, -R, and -C (which set filtertype)
This could be - -filtertype={vlf, lfo, resp, card, none), default is none.  The only thing is that using the -F option can’t fit into that structure, but that’s maybe ok.,
> 
> -h (to set the histogram length) is duplicated with the flag to show the usage. I was thinking it could be replaced with —histlen.
Agreed
> The variable check_autocorrelation is set to True by several arguments, but it doesn't look like it can ever be False, so I think it could be removed.
That’s left over from when it was False by default and set by a command line option (since removed).  Probably no reason for there to be a way to disable it, since it’s quick and it doesn’t hurt anything to do it even if it doesn’t get used.  Then --accheck goes away as well.
> The variable shiftall also seems to always be True.
That’s used for internal debugging.  I like to have the option to turn that on and off for testing.

While you’re at it, we could change the following:
-T -> —savelags
-b -> —bipolar
-c -> —cifti
-M -> —globalmean
-m -> —meanscale
-I -> —invert
-o -> —offsettime=
-a option can be removed and antialias left as true - there’s no good reason a user would ever turn it off (I still want the internal variable for debugging)
-O -> —oversampfac=
-N -> —numnull=
-b -> —butterorder=ORDER  turns butterworth filter on and sets filter order to ORDER

**Comment by bbfrederick** [comment]
Now that I've had a night to sleep on it, I think while this is the right direction, there are a few old style options that are used frequently that I think shouldn't just go away without a transition.  Specifically, -V, -L, -R, -C (or maybe just -L, since that's the one that is almost always used - the others are rare) should continue to exist with a deprecation warning in the comments.  Also, -M and -N are fairly common.  All the others are so rarely used that if anybody tries them, they'll see the error and fix it.  BTW, -t has had a deprecation warning for a few months now, so I think it can go away now.

**Comment by tsalo** [comment]
This is great, thank you! I'll probably have some more questions before I open a WIP PR, but I think I should have an operational parser ready for review soon.

**Comment by tsalo** [comment]
Sorry for the delay. I haven't had a chance to update the parser I wrote [here](https://github.com/tsalo/rapidtide/blob/0fae68774d704228bc4e873e9f0fda5b81d6185c/rapidtide/parse.py) with recent changes to the code, but that can always be done later. Here is an example argparse-generated help output ([help.txt](https://github.com/bbfrederick/rapidtide/files/2141777/help.txt)).
WDYT?

Also, regarding the structure of the Python package, I think that the `tide_funcs.py` file (which has ~3k lines) may be a little unwieldy. Do you have any preferences for how that could be broken down into submodules? You've included useful section headers, like [this one](https://github.com/tsalo/rapidtide/blob/0fae68774d704228bc4e873e9f0fda5b81d6185c/rapidtide/tide_funcs.py#L2982). Would those sections work as separate submodules (e.g., `utils.py`)?

**Comment by bbfrederick** [comment]
That looks quite good to me - it's less compact, but it looks a lot more maintainable.  A couple things:

There are additional rarely used arguments that should be converted to more easily interpreted double hyphen arguments that I hadn't noticed until reading your code.  -S should be --simulate, -p should be --prewhiten, -P should be --saveprewhiten (which would enable prewhiten), -A should just go away, -f should also have --spatialfilt, -s should just become --sigmalimit.

In order to nudge people towards more informative arguments, we should add --lfo to -L, --cardiac to -C, --resp to -R, --vlf to -V, and --arb to -f.

Re: splitting up tide_funcs, yes, that's a good idea.  That's kind of major surgery, so I feel like I should probably do that.  I've started the process by pulling all the I/O functions into io_funcs.py (I'm not sure about the naming convention - maybe io.py would be better?  I don't feel strongly about it.). I'll look over tide_funcs and see what other natural modules there are (as you say, utils is a good one too, just have to decide what is a utility).

**Comment by bbfrederick** [comment]
Oh, and I think you're right about usetmask being set to true if tmaskname is not None.

**Comment by tsalo** [comment]
Thanks for catching those arguments. I've updated my fork with the corrected parser and 
[here](https://github.com/bbfrederick/rapidtide/files/2149797/help.txt) is the updated help.

Here were some ideas I had for splitting up `tide_funcs.py`:
- The `version` function can probably be dropped because you have `version.py` and that info can be loaded automatically in `setup.py`.
- The remaining sections could probably be divided like so:
    - “timing functions” --> `rapidtide.timing`
    - “histogram functions” --> `rapidtide.histogram`
    - “probability functions” --> `rapidtide.probability`
    - “correlation functions” --> `rapidtide.correlation`
        - There’s actually a lot here, so this could be broken down further.
    - “fitting functions” --> `rapidtide.fit`
    - “resampling and time shifting functions” --> `rapidtide.resample` and `rapidtide.shift`
    - “window functions” --> `rapidtide.window`
    - “normalization functions” --> `rapidtide.normalize` or `rapidtide.norm`
    - “filtering functions” --> `rapidtide.filter`
    - “spectral analysis functions” --> `rapidtide.spectrum`
    - “utility functions” --> `rapidtide.utils`
- Additional layers of submodules could be added anywhere to simplify the architecture (e.g., `rapidtide.probability` + `rapidtide.correlation` --> `rapidtide.stats.probability` + `rapidtide.stats.correlation`
- Regarding naming the submodules, I think `_funcs` can be redundant. Of course, for functions that are only used internally, that doesn’t really matter. It’s only an issue if users would call the submodule.
- Some of the scripts could be incorporated into the package itself as workflows. For example, rapidtide2 could be a workflow in `rapidtide.workflows`, and then the CLI script `rapidtide2` would essentially just contain the argument parsing and a call to `rapidtide.workflows.rapidtide2`. That’s the approach we’ve taken in [tedana](https://github.com/ME-ICA/tedana), and I think it’s pretty good.

Do you think that breakdown would work? I am not familiar enough with rapidtide's code to know if there is anything that would make it difficult to split `tide_funcs` into so many submodules (beyond the fact that it's a lot of work), so it could be that this architecture is completely inappropriate.

**Comment by bbfrederick** [comment]
I've started a new issue re: splitting tide_funcs, since this seems like an important issue in its own right.

**Comment by bbfrederick** [comment]
Just in case you missed the other issue (I don't know how to tag you in an issue post), I've now split up tide_funcs.py in much the way you suggested.  I kept windowing and filtering in the same module (filter.py), kept resampling and shifting together (resample.py), kept probability and histograms together (stats.py), and put timing, normalization and spectral stuff in miscmath.py.  I also split all I/O functions into io.py.  fit.py, correlate.py, and util.py round out the set.  It was a bit painful, and there may be a few calls I missed that will shake out in the next few weeks, but it was a very good idea, and I think things are much cleaner as a result.  The eventual split was to balance the need for some modules to call others, and resolving circular dependencies.

**Comment by tsalo** [comment]
Oops, sorry. I'll jump over to that issue to discuss the package organization. It looks really awesome, and should be useable in Python at this point too.

Regarding argparse, I've started trying something on my fork that I think works quite well. The parser I worked on (or something like it) can go in `scripts` (like [this](https://github.com/tsalo/rapidtide/blob/docs/rapidtide/scripts/run_rapidtide.py)), while the actual pipeline could go in a new submodule called `workflows` (like [this](https://github.com/tsalo/rapidtide/blob/docs/rapidtide/workflows/rapidtide2.py)). The parser would just take in command line arguments and parse them, before sending them to the main function associated with whichever pipeline in `workflows`, which would do all of the actual work. This allows users to call each full pipeline in pure Python, without preventing command line usage. It also makes it easier to update the documentation pages in tandem with changes to the code, because RTD will do that automatically. The only hard part will be making sure that changes to the input variables in the `workflows` function are also updated in the CLI (and vice versa).

[Here](https://github.com/bbfrederick/rapidtide/files/2195271/html.zip) is what the html files could look like with this setup. The Usage page renders the argparse-based parser automatically, while the API page will include documentation for each specified functions or class, including the workflow function called by the CLI. Once detailed numpydoc-style docstrings are added for the functions, the API page will look a lot better. Also I just used the first 1-2 functions or classes from each module to fill out the API page to start.

**Comment by bbfrederick** [comment]
Hey, sorry for the delay - was at a conference all last week and didn't really have a lot of time for software stuff.

I'm all on board with most of this - I like the cleaner argument parsing and self-documentation (I REALLY have to start writing docstrings.  Not that much of an issue when the codebase was small, but even I have to keep going back to the source to figure out what the arguments are), but I'm still a little iffy on the workflows part.  I like that idea, but I do worry about the issue you raised about keeping the function call and the argparse in sync.  If we do want to go down that route, I'd invert the naming convention.  Instead of "run_rapidtide2.py" for the call script, I'd call that "rapidtide2" (or rapidtide2.py - I'm not a big fan of the .py suffix convention for executables, but maybe that's just me.  I don't think executables need to tell you what language they are in.  It was never an issue in C, because the source and executable were separate files).  Then the workflow would be "run_rapidtide2.py".  Or does that mess up the automatic documentation part?

**Comment by bbfrederick** [comment]
Ok, I've made docstrings for filter.py (and done a lot of systematizing and cleaning of the code to make arguments more consistent).  I'll try to get to the other modules as time permits.

**Comment by tsalo** [comment]
The executable could definitely stay `rapidtide`, while the workflow could just be `rapidtide.workflows.rapidtide` or `rapidtide.workflows.run_rapidtide`. That won't cause any problems with the automated documentation, although the actual executable associated with the Python script ends up being a separate file, from what I can tell. For example, in another package I've worked on, `tedana`, the CLI scripts are named run_t2smap.py and run_tedana.py, but when I type `which t2smap` or `which tedana` in my terminal, it points to files in `python/bin/` named t2smap (no .py) and tedana. You can see how the setup.py file is configured for the CLI [here](https://github.com/ME-ICA/tedana/blob/master/setup.py#L43-L45).

Probably the better approach is to support the workflows in Python, but in a single file. Each of the scripts would have to go in a workflows module (as is sort of the case in my original proposal), without a scripts folder, and would have the .py suffix. Both the actual workflow and the argument parser would be in the same file, but the CLI would be configured in setup.py (as with tedana) and would generate the executables named after the workflows that you want.

The docstrings look great. There are a few things that are incompatible with the Sphinx scheme I was setting up, but I can open a pull request that addresses them and adjusts how Sphinx is configured for the API documentation. Is that alright?

**Comment by bbfrederick** [comment]
Ok, that organization seems reasonable (BTW, tedana looks very interesting - thanks for the pointer!)

Re: docstrings - I don't have any preferred way of doing them, so I'm willing to change to some other style/formatting if that will make them more compatible - I was just using a style guide from a random website.  What specifically doesn't work with your Sphinx scheme?

**Comment by tsalo** [comment]
Awesome! I realize that my parser for rapidtide2 is probably out of date at this point, but I'd be happy to write a prototype of the workflows approach with one of your simpler scripts. Are there any in particular that would be good for this?

The only things that are incompatible are 1) the `:param [paramname]:` at the end of each parameter definition is unnecessary and 2) mentioning the function name in the docstring is redundant because the description renders next to the function name on the site. You can see the first issue in the first image below and the second in the second.

-------------
<img width="600" alt="screen shot 2018-07-26 at 4 05 38 pm" src="https://user-images.githubusercontent.com/8228902/43285741-eba36642-90ed-11e8-947f-ef6cadbdfadf.png">

-------------
<img width="600" alt="screen shot 2018-07-26 at 4 05 48 pm" src="https://user-images.githubusercontent.com/8228902/43285745-ee831678-90ed-11e8-9009-c8a005022606.png">

-------------
There are also a couple of good shortcuts that aren't necessary, but might be nice to use:
- For parameters with a limited set of possible values, instead of defining the type as `str` and listed the options in the description, you can just do `param_name : {'option1', 'option2', 'option3'}` according to numpydoc convention.
- It's hard for me to tell without better knowledge of the filter functions and their common usage, but it looks like you used the section `Other Parameters` for all keyword arguments, although it's meant for the ones that are uncommonly used and generally you want to use it only if there are a large number of keyword arguments (as in the workflows). For keyword parameters included in the `Parameters` section, you can add in `, optional` after the variable type (e.g., `param_name : str, optional`). 
- If you use [intersphinx](http://www.sphinx-doc.org/es/stable/ext/intersphinx.html), then mentions of parameters and outputs that are objects from other packages (i.e., numpy arrays) will link to the package's documentation for those objects. It will also link within the rapidtide package documentation if classes are fed through functions.

So instead of 
```
Parameters
----------
inputdata : 1D numpy array
    Input data to be filtered
```
you would have
```
Parameters
----------
inputdata : 1D :obj:`numpy.ndarray`
    Input data to be filtered
```

**Comment by tsalo** [comment]
I've opened a PR with the proposed changes to the documentation site (#6). Please feel free to request any changes you want and I'll be happy to make them to the PR. The PR starts work on the workflows submodule but will only affect the RTD site because I've left `scripts/` untouched.

**Comment by bbfrederick** [comment]
Responses roughly in order of your messages:

If you wanted to try a prototype for the argparse workflow, showxcorr (or even better, showxcorrx) might be a good place to start, since they have a rich set of options, but I'm currently not touching them, so they won't change out from under you...

Re: the :param variablename: thing - I put those in because PyCharm complains when I don't.  So if you could update the sphinx processor to recognize them, that might be good.

I will switch to the param_name convention you suggest.

In the two functions you showed, I presume that other than the param weirdness, the ssmooth one is the desired format.  It looks better to me at least.  I'll take the function names out btw.

Now as to the PR - I have no idea how to process one.  You mention making changes - do I just send you messages to tell you what I want to change?  Do I have a way of testing it out?  Sorry, total noob...

**Comment by bbfrederick** [comment]
Argh!  I think I screwed up.  I went through and approved your changes, and I thought I was merging them into the main branch, but it looks like maybe I deleted them?  They seem to be gone...

**Comment by tsalo** [comment]
Sorry, I know the pull request system is a little weird. Thanks for merging it though!

I will put together the `showxcorrx` workflows at some point in the near future and can open another PR when that's ready. I'll also try to figure out the PyCharm/Sphinx issue. I _think_ PyCharm is set up for a specific kind of docstring format (not numpydoc, which I've been using for RTD), and that could be why it wants that particular scheme. That said, I've never actually used PyCharm, so I'm not 100% sure about that yet.

It looks like the docs are failing because rapidtide isn't installed (see [here](https://readthedocs.org/projects/rapidtide/builds/7552593/)). I think it could be something in your RTD settings. If you go to the Admin: Advanced Settings tab on the RTD [project site](https://readthedocs.org/projects/rapidtide/), is **Install Project** turned on? You can see what the page looks like below (though for a different project I've built on RTD):

<img width="600" alt="screen shot 2018-07-27 at 11 47 41 am" src="https://user-images.githubusercontent.com/8228902/43331741-6b507802-9194-11e8-9112-30ac4b5d97e2.png">

**Comment by bbfrederick** [comment]
Install project was not on - it is now.  If it doesn't start doing something soon I'll try kicking it.

**Comment by bbfrederick** [comment]
Weird.  It seems to want to install rapidtide to run the docs processing. It's failing because pyqt4 is not installed.

**Comment by bbfrederick** [comment]
Fixed that by commenting that out of the __init__.py file (don't know why it was there to begin with.  Now it complains that it can't find sphinxarg.ext.  Is this something of yours?

**Comment by tsalo** [comment]
That one's my fault, but I'm not sure why RTD isn't installing it. It's in the doc requirements in setup.py [here](https://github.com/bbfrederick/rapidtide/blob/master/setup.py#L184). My guess is that it needs to go in [requirements.txt](https://github.com/bbfrederick/rapidtide/blob/master/requirements.txt) (along with numpydoc and m2r). I can open a fix PR or you could just update the file directly?

**Comment by bbfrederick** [comment]
I updated it.  Waiting to see if it worked - the version is locked for 5 minutes because I updated it too much too quickly...

**Comment by tsalo** [comment]
Sorry, `sphinxarg.ext` should be `sphinx-argparse`. Using a docs-specific requirements file was a great idea.

**Comment by bbfrederick** [comment]
Wow - that looks great!  Inspiration to do more docstrings...

I think I messed up the main index.rst page.  I made a tiny change, and couldn't figure out how to resolve the conflict.  Would just dropping in the one from your branch work?

**Comment by tsalo** [comment]
When you cleaned up docs/index.rst to remove extra introduction text, you deleted a few lines that were necessary for Sphinx:
```
Contents
========
.. toctree::
   :maxdepth: 2
```

which should go after
```
Citing rapidtide
-------------------

  Frederick, B, rapidtide [Computer Software] (2016). Available from
  https://github.com/bbfrederick/rapidtide. doi:`10.5281/zenodo.814990`_

.. _10.5281/zenodo.814990: https://doi.org/10.5281/zenodo.814990
```

but before

```
   introduction
   usage
   api
   contributing
   whats_new
```

If you add those lines back in it should work I think.

**Comment by tsalo** [comment]
It looks good now! WDYT?

**Comment by bbfrederick** [comment]
Added the lines you suggested, then added a blank line between introduction and the line before it…

**Comment by bbfrederick** [comment]
Ok, new issue.  I've added lots of docstrings, but the documentation pages never change for anything other than filter.py.  Is there a file 'function.rst' that you have that didn't make it into to docs directory?  sphinx-autopen seems to be looking for it.

**Comment by tsalo** [comment]
In order to show functions or classes in the API page, you need to update [api.rst](https://github.com/bbfrederick/rapidtide/blob/master/docs/api.rst) with the function names manually. This can be tedious, but on the bright side it's a good opportunity to decide which functions should be public and which (if any) should be private.

You also probably want to get rid of the following, because rapidtide.filter isn't a class.

```
.. autoclass:: rapidtide.filter
    :no-members:
```

**Comment by bbfrederick** [comment]
Accepted the PR.  I really like the example!  

I'm sold on the workflow idea - I'm seeing a lot of advantages.  Two that come to mind right away are to make routine testing easy (while I've thought about testing, I really haven't made any progress towards it).  Also, since rapidtide2 and rapidtide2x share so much code, having a merged workflow with multiple entry points would make it much easier to keep things in sync.

**Comment by tsalo** [comment]
I agree that the workflows will make integration testing much easier, as you won't have to rely on `subprocess` to call the workflows from the terminal.

I can definitely update the change log.

Given that you support the workflows approach, which involves both argparse-based argument parsing for the command line scripts and within-Python usage for the workflows, I think it makes sense to close this issue. I can open a new issue to track progress on translating the scripts over to workflows and maybe link it to a future release/milestone?

**Comment by bbfrederick** [comment]
Yes, let's close this, and move on to other better things.
